keps/sig-node/xxx-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1368 @@
+  directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
+  to use part of the shareable device inside a container and other
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow to allocate part of the device.
+
+- *Optional allocation*: When deploying workload I’d like to specify
+  soft(optional) device requirements. If the device exists and it’s
+  allocatable it would be allocated. If not - workload will be run on
+  the node without a device. GPU and crypto-offload engines are
+  examples of this kind of device. If they’re not available, workloads
+  can still run by falling back to using only the CPU for the same
+  task.
+
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
+- *Support Over the Fabric devices*: When deploying a container, I’d
+  like to utilize devices available over the Fabric (PCIe, CXL,
+  Network, special links, etc).
+
+  *Limitation*: Device Plugins framework supports only local devices.
+
+- *Access to the container image*: When deploying a container that
+  needs access to a GPU device, I would like my container to
+  gracefully fail to start (rather than SIGSEGV at runtime) if a
+  minimum driver requirement is not met on the host.
+
+  *Limitation*: GPU driver requirements are currently stored as labels
+  on the container image, and device plugins do not have access to the
+  container image.
+
+- *Perform container runtime specific operations*: When deploying a container
+  that needs access to a device, I would like to be able to reuse the
+  same pod spec, irrespective of the underlying container runtime in
+  use (e.g. kata vs. runc).
+
+  *Limitation*: At present, there is no way to perform
+  runtime-specific operations that may need to occur as a result of
+  injecting a device into a container (device plugins are runtime
+  agnostic). A good example is supporting GPU passthrough
+  virtualization on kata vs. runc.
+
+- *Access to the plugin container*: When deploying a device plugin, I
+  would like to ensure that all of the operations that need to occur
+  as part of “injecting” a device into a container, also occur for the
+  “plugin container” itself.
+
+  *Limitation*: At present, there is a chicken-and-egg problem to
+  supporting this because the device plugin is the sole dictator of
+  which containers will have these operations performed on them.
```

Comment:
It is not clear to me what is meant by "all of the operations that need to occur  as part of “injecting” a device into a container, also occur for the  “plugin container” itself". Do you mean the same operations need to be applied to the plugin and the container?

Reply:
I copied the motivation section from an earlier document written by @kad and raised the same question myself, without getting it resolved before moving the text here.

@kad: can you clarify what was meant here?

Reply:
@cathyhongzhang @pohly "all operations" means in that scope preparation of the device itself (e.g. FPGA programming before it can be used) or inspecting variables, container meta-data (e.g. annotations) to select right set of mounts (e.g. libraries in NVIDIA case). 

Same goes to monitoring tasks: e.g. if you have "device plugin" daemon set, for some devices to properly detect parameters or monitoring health of devices, DaemonSet should have access to control devices and be able to perform e.g. `ioctl` calls to it. So, it might be scenarios where first class of devices is used just with simple "detect over PCI ID" to be able to start real node agent that would have access to all control device nodes to properly expose device internal state/resoruces. (that example was mentioned by NVIDIA in early conversations).

Reply:
From the perspective of the current NVIDIA implementation, the requirements for the plugin container and an (application) container could be handled differently. For the purpose of the plugin, it is sufficient to know which devices are available and what their properties are. In an environment where configurations are not expected to change dynamically, a CDI spec (for example) could be configured once and consumed by the plugin.

Even without the CDI spec, starting the plugin container as a privileged container and including a small number of libraries as mounts would allow for the required functionality.

I think the key is whether selection by the plugin container is considered "use" of a device. The answer may be vendor / implementation specific.